apis/v1alpha1/httproute_types.go

```diff
@@ -328,79 +328,94 @@ type HTTPRouteMatch struct {
+	//   are encouraged to support extended filters.
+	//
+	// - Custom: Filters that are defined and supported by specific vendors.
 	//   In the future, filters showing convergence in behavior across multiple
 	//   implementations will be considered for inclusion in extended or core
-	//   conformance rings. Filter-specific configuration for such filters
+	//   conformance levels. Filter-specific configuration for such filters
 	//   is specified using the ExtensionRef field. `Type` should be set to
-	//   "ImplementationSpecific" for custom filters.
+	//   "Custom" for custom filters.
 	//
-	// Implementers are encouraged to define custom implementation
-	// types to extend the core API with implementation-specific behavior.
+	// Implementers are encouraged to define custom implementation types to
+	// extend the core API with implementation-specific behavior.
 	//
 	// +unionDiscriminator
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:validation:MaxLength=100
-	Type string `json:"type"`
+	Type HTTPRouteFilterType `json:"type,omitempty"`
 
-	// ExtensionRef is an optional, implementation-specific extension to the
-	// "filter" behavior.  The resource may be "configmap" (use the empty
-	// string for the group) or an implementation-defined resource (for
-	// example, resource "myroutefilters" in group "networking.acme.io").
-	// Omitting or specifying the empty string for both the resource and
-	// group indicates that the resource is "configmaps".
-	// ExtensionRef MUST NOT be used for core and extended filters.
+	// RequestHeaderModifier defines a schema for a filter that modifies request
+	// headers.
+	//
+	// Support: Core
+	//
 	// +optional
-	ExtensionRef *LocalObjectReference `json:"extensionRef,omitempty"`
-
-	// Filter-specific configuration definitions for core and extended filters
+	RequestHeaderModifier *HTTPRequestHeaderFilter `json:"requestHeaderModifier,omitempty"`
 
+	// RequestMirror defines a schema for a filter that mirrors requests.
+	//
+	// Support: Extended
+	//
 	// +optional
-	RequestHeader *HTTPRequestHeaderFilter `json:"requestHeader,omitempty"`
+	RequestMirror *HTTPRequestMirrorFilter `json:"requestMirror,omitempty"`
 
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "filter" behavior.  The resource may be "configmap" or an
+	// implementation-defined resource (for example, resource "myroutefilters"
+	// in group "networking.acme.io"). ExtensionRef MUST NOT be used for core
+	// and extended filters.
+	//
+	// Support: Implementation-specific
+	//
 	// +optional
-	RequestMirror *HTTPRequestMirrorFilter `json:"requestMirror,omitempty"`
+	ExtensionRef *LocalObjectReference `json:"extensionRef,omitempty"`
 }
 
-// HTTPRequestHeaderFilter defines configuration for the
-// RequestHeader filter.
+// HTTPRouteFilterType identifies a type of HTTPRoute filter.
+// +kubebuilder:validation:Enum=RequestHeaderModifier;RequestMirror;Custom
+type HTTPRouteFilterType string
```

Comment:
Question: should this instead be called RouteFilterType?

Two reasons:
1. We could keep doors open for filters on other protocols. Unlikely and I'm myself not convinced if we will ever get there.
2. Some implementations might chose to also use these filters in resources other than HTTPRoute.

Reply:
Although filters may be common across the protocol stack, they will have different meaning so I figured they should be protocol-specific. The following is an example of the request header modifier for `TCPRoute`:
```
// TCPRouteRule is the configuration for a given rule.
type TCPRouteRule struct {
...
	// Filters define the filters that are applied to requests that match
	// this rule.
	Filters []TCPRouteFilter `json:"filters,omitempty"`
}

// TCPRouteFilterType identifies a type of TCPRoute filter.
type TCPRouteFilterType string

const (
	// RequestHeaderModifierTCPRouteFilter can be used to modify the header of
        // a TCP segment before it is sent to the upstream target.
	//
	// Support: Core
	RequestHeaderModifierTCPRouteFilter TCPRouteFilterType = "RequestHeaderModifier"
...
)
```

Thoughts?

Reply:
I don't see any harm in doing that. It leads to having more types in the API, but then I could also argue that separating out types is better in this case. Let's leave this be.